components/tidb_query/src/codec/row/v2/decoder.rs

```diff
@@ -0,0 +1,332 @@
+                let start = if idx > 0 { offsets[idx - 1] } else { 0usize };
+                row.insert(*id, (&values[start..*offset]).decode_datum(ctx, tp)?);
+            } else if null_ids.binary_search(id).is_ok() {
+                row.insert(*id, Datum::Null);
+            } else {
+                row.insert(*id, tp.get_default_val().decode_datum(ctx, tp)?);
+            }
+        }
+        Ok(row)
+    }
+}
+
+impl<T: BufferReader> RowDecoder for T {}
+
+trait DatumDecoder: NumberDecoder + DecimalDecoder + JsonDecoder {
+    #[inline]
+    fn decode_datum(&mut self, ctx: &EvalContext, ft: &dyn FieldTypeAccessor) -> Result<Datum> {
+        use core::convert::TryInto;
+        use tidb_query_datatype::FieldTypeTp::*;
+        match ft.tp() {
+            Tiny | Short | Long | LongLong | Int24 | Year => {
+                if ft.is_unsigned() {
+                    Ok(Datum::U64(self.decode_u64()))
+                } else {
+                    Ok(Datum::I64(self.decode_i64()))
+                }
+            }
+            VarChar | VarString | String | TinyBlob | MediumBlob | LongBlob | Blob => {
+                Ok(Datum::Bytes(self.bytes().to_vec()))
+            }
+            Double | Float => Ok(Datum::F64(f64::from_bits(self.decode_u64()))),
+            Date | DateTime | Timestamp => Time::from_packed_u64(
+                self.decode_u64(),
+                ft.tp().try_into().unwrap(),
+                ft.decimal() as i8,
+                &ctx.cfg.tz,
+            )
+            .map(Datum::Time),
+            NewDecimal => self.decode_decimal().map(Datum::Dec),
+            JSON => self.decode_json().map(Datum::Json),
+            Duration => {
+                mysql::Duration::from_nanos(self.decode_i64(), ft.decimal() as i8).map(Datum::Dur)
+            }
+            tp => return Err(invalid_type!("unsupported data type `{:?}`", tp)),
+        }
+    }
+
+    #[inline]
+    fn decode_u64(&mut self) -> u64 {
+        let data = self.bytes();
+        match data.len() {
+            1 => u64::from(data[0]),
+            2 => u64::from(NumberCodec::decode_u16_le(data)),
+            4 => u64::from(NumberCodec::decode_u32_le(data)),
+            _ => NumberCodec::decode_u64_le(data),
+        }
+    }
+
+    #[inline]
+    fn decode_i64(&mut self) -> i64 {
+        let data = self.bytes();
+        match data.len() {
+            1 => i64::from(data[0]),
+            2 => i64::from(NumberCodec::decode_i16_le(data)),
+            4 => i64::from(NumberCodec::decode_i32_le(data)),
+            _ => NumberCodec::decode_i64_le(data),
+        }
+    }
+}
+
+impl<T: BufferReader> DatumDecoder for T {}
+
+// convert number to usize or i64
+// ids need to convert i64 since the column id is i64
+// offsets need to convert of usize, since it will used as range index later
+pub trait Converter {
+    fn to_usize(&self) -> usize;
+    fn to_i64(&self) -> i64;
+}
+
+impl Converter for u8 {
```

Comment:
Prefer:
```rust
impl<T: num_traits::PrimInt> for T {
    fn to_usize(&self) -> usize {
        self.to_le() as usize
    }

    fn to_i64(&self) -> i64 {
        i64::from(*self)
    }
}
```
instead of implement `Converter` individually for `u8/u16/u32`.